producers/prometheus/prometheus.go

```diff
@@ -15,17 +15,261 @@
+	// TODO (philipnrmn): rewrite this section to avoid use of the store
+	go func() {
+		promLog.Debug("Prom producer listening for incoming messages on metricsChan")
+		for {
+			// read messages off the channel,
+			// and give them a unique name in the store
+			message := <-p.metricsChan
+
+			var name string
+			switch message.Name {
+			case producers.NodeMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.MesosID)
+
+			case producers.ContainerMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.ContainerID)
+
+			case producers.AppMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.ContainerID)
+			}
+
+			for _, d := range message.Datapoints {
+				p.writeObjectToStore(d, message, name)
+			}
+		}
+	}()
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(p)
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(
+		registry, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
+
+	server := &http.Server{
+		Addr:    net.JoinHostPort("localhost", strconv.Itoa(p.config.Port)),
+		Handler: mux,
+	}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			if err != http.ErrServerClosed {
+				promLog.Printf("E! Error creating prometheus metric endpoint, err: %s\n",
+					err.Error())
+			}
+		}
+	}()
 	return nil
 }
+
+// Describe passes all the available stat descriptions to Prometheus; we
+// send a single 'dummy' stat description, and then create our actual
+// descriptions on the fly in Collect() below.
+func (p *promProducer) Describe(ch chan<- *prometheus.Desc) {
+	promLog.Warn("Describe() was called")
+	prometheus.NewGauge(prometheus.GaugeOpts{Name: "Dummy", Help: "Dummy"}).Describe(ch)
+}
+
+// Collect iterates over all the metrics available in the store, converting
+// them to prometheus.Metric and passing them into the prometheus producer
+// channel, where they will be served to consumers.
+func (p *promProducer) Collect(ch chan<- prometheus.Metric) {
+	for _, obj := range p.store.Objects() {
+		message, ok := obj.(producers.MetricsMessage)
+		if !ok {
+			promLog.Warnf("Unsupported message type %v", obj)
+			continue
+		}
+		dims := dimsToMap(message.Dimensions)
+
+		for _, d := range message.Datapoints {
+			promLog.Debugf("Processing datapoint %s", d.Name)
+			var tagKeys []string
+			var tagVals []string
+			for k, v := range dims {
+				tagKeys = append(tagKeys, sanitizeName(k))
+				tagVals = append(tagVals, v)
+			}
+			for k, v := range d.Tags {
+				tagKeys = append(tagKeys, sanitizeName(k))
+				tagVals = append(tagVals, v)
+			}
+
+			name := sanitizeName(d.Name)
+			val, err := coerceToFloat(d.Value)
+			if err != nil {
+				promLog.Warnf("Bad datapoint value %q: %s", d.Value, err)
+				continue
+			}
+			desc := prometheus.NewDesc(name, "DC/OS Metrics Datapoint", tagKeys, nil)
+			metric, err := prometheus.NewConstMetric(desc, prometheus.GaugeValue, val, tagVals...)
```

Comment:
Are we setting the timestamp on the `metric` anywhere? if our store holds more than one time series for a particular metric this could cause overwrites if we're not doing this.

Reply:
Our store can't hold two metrics with the same name and different timestamps - the older one will be replaced by the newer one. The lack of timestamps seemed odd to me too, but it's what Prometheus recommend in their exporter documentation:

> ..scrapes should be synchronous. Accordingly, you should not set timestamps on the metrics you expose, let Prometheus take care of that. If you think you need timestamps, then you probably need the Pushgateway instead.

(from https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling)

Reply:
aha, ok makes sense. thanks for the clarification 👍